lib/kafka/sasl/oauth.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module Kafka
+  module Sasl
+    class OAuth
+      OAUTH_IDENT = "OAUTHBEARER"
+
+      # token_provider: THE FOLLOWING INTERFACE MUST BE FULFILLED:
+      #
+      # [REQUIRED] TokenProvider#token      - Returns an ID/Access Token to be sent to the Kafka client.
+      #   The implementation should ensure token reuse so that multiple calls at connect time do not
+      #   create multiple tokens. The implementation should also periodically refresh the token in
+      #   order to guarantee that each call returns an unexpired token. A timeout error should
+      #   be returned after a short period of inactivity so that the broker can log debugging
+      #   info and retry.
+      #
+      # [OPTIONAL] TokenProvider#extensions - Returns a map of key-value pairs that can be sent with the
+      #   SASL/OAUTHBEARER initial client response. If not provided, the values are ignored. This feature
+      #   is only available in Kafka >= 2.1.0.
+      #
+      def initialize(logger:, token_provider:)
+        @logger = TaggedLogger.new(logger)
+        @token_provider = token_provider
+      end
+
+      def ident
+        OAUTH_IDENT
+      end
+
+      def configured?
+        @token_provider
+      end
+
+      def authenticate!(host, encoder, decoder)
+        # Send SASLOauthBearerClientResponse with token
+        @logger.debug "Authenticating to #{host} with SASL #{OAUTH_IDENT}"
+
+        encoder.write_bytes(initial_client_response)
+
+        begin
+          # receive SASL OAuthBearer Server Response
+          msg = decoder.bytes
+          raise Kafka::Error, "SASL #{OAUTH_IDENT} authentication failed: unknown error" unless msg
```

Comment:
Reading through the broker code, it looks like the `msg` they return is meant to be a JSON object describing the error.
I wonder if it's worth putting the object either into the exception object, or logging it, or something?

I don't think we can interpret the failure, but perhaps the client author could?

Reply:
`decoder.bytes` will raise an error and crash if we get a bad response from Kafka and it will just log a general error in the `rescue` block. 

I'm guessing it assumes that if we receive a msg at all then Kafka succeeded, otherwise it'll internally decide if the response is bad. 

It probably does make sense to get the actual response from Kafka and log a more meaningful error, since Kafka will return a proper code, but that's probably for another PR.

Reply:
That's alright by me. It looks like all of the SASL implementations are making a similar assumption, so 🤷‍♂️

Reply:
https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-ErrorCodes 

This might be useful for a future PR?